### PR TITLE
Add XGMI Info to RSMI plugin

### PR DESCRIPTION
**Pull request by miketxli**
Change-Id: I95524183adee7730a53ec505505045e480a75132
Signed-off-by: Mike Li <Tianxinmike.Li@amd.com>

### CODE REVIEW

**Comment by bgoglin** [comment]
Hello. Can you clarify what information is added here for an example of platform, and if you plan to add more in the future?

**Comment by miketxli** [comment]
With this change I have added the information regarding the XGMI connection of the GPUs for RSMI plugin. If GUP's are connected through XGMI, it will show the HIVE ID and the GPU's connected to this each GPU. The following is an example with 2 GPU connected with XGMI:
          PCIBridge L#6 (busid=0000:3c:04.0 id=10b5:9797 class=0604(PCIBridge) link=15.75GB/s buses=0000:[3d-3f])
            PCIBridge L#7 (busid=0000:3d:00.0 id=1002:14a0 class=0604(PCIBridge) link=15.75GB/s buses=0000:[3e-3f])
              PCIBridge L#8 (busid=0000:3e:00.0 id=1002:14a1 class=0604(PCIBridge) link=31.51GB/s buses=0000:[3f-3f])
                PCI L#3 (busid=0000:3f:00.0 id=1002:66a1 class=0380(Display) link=15.00GB/s)
                  GPU L#1 (Backend=RSMI GPUVendor=AMD GPUModel="Vega 20" AMDSerial=PCB026064-0052 AMDUUID=61f2504172da5ee8 XGMIHIVEID=b7f9f50f7508bf17 XGMILinks="rsmi1 ") "rsmi0"
          PCIBridge L#9 (busid=0000:3c:0c.0 id=10b5:9797 class=0604(PCIBridge) link=15.75GB/s buses=0000:[41-43])
            PCIBridge L#10 (busid=0000:41:00.0 id=1002:14a0 class=0604(PCIBridge) link=15.75GB/s buses=0000:[42-43])
              PCIBridge L#11 (busid=0000:42:00.0 id=1002:14a1 class=0604(PCIBridge) link=31.51GB/s buses=0000:[43-43])
                PCI L#4 (busid=0000:43:00.0 id=1002:66a1 class=0380(Display) link=15.00GB/s)
                  GPU L#2 (Backend=RSMI GPUVendor=AMD GPUModel="Vega 20" AMDSerial=PCB026064-0121 AMDUUID=c3c4606172dc76b7 XGMIHIVEID=b7f9f50f7508bf17 XGMILinks="rsmi0 ") "rsmi1"

I'm planing to add GlobalMemorySize, ComputeUnits, and CacheSize in next step, since it require some changes from ROCm-SMI library.

**Comment by bgoglin** [comment]
What's a hive?
Can you have things like a switch between multiple GPUs? Or are all these links only between GPUs? Is it possible to have multiple XGMI links between same GPUs?

**Comment by miketxli** [comment]
A Hive is group of GPU's connected together. It's similar like a switch between 2 or 4 GPU's. Within the group, they have same hive id.  Only one link for same GPUs.  links are only between GPU's. It may extent to CPU in the future.

**Comment by bgoglin** [comment]
Sorry for the delay, I was on vacation. And thanks for the info.
If I have a hive with 3 GPUs, should XGMILinks contain two rsmiX names separated by a space? It seems to me your code would just keep the last one.
Also wouldn't XGMIPeers be better than XGMILinks for the info name?
Do you have some info about the XGMI link speed/width? Either from the API or hardwired from the hardware doc. It's something we'll put in the distance matrix (next thing on my TODO list).

**Comment by miketxli** [comment]
If I have a hive with 3 GPUs, should XGMILinks contain two rsmiX names separated by a space? It seems to me your code would just keep the last one.
That's a bug, I will fix it.

Also wouldn't XGMIPeers be better than XGMILinks for the info name?
Ok

Do you have some info about the XGMI link speed/width? Either from the API or hardwired from the hardware doc. It's something we'll put in the distance matrix (next thing on my TODO list).
Vega 20 has 2 Infinity Fabric(tm) links per chip, each supports a bidirectional bandwidth of up to 100 GB/s (1x 25 Gbps x16 link upstream and 1x 25 Gbps x16 link downstream). That's from hardware and is a theoretical numbers.

File: hwloc/topology-rsmi.c [hunk 1]
```diff
@@ -112,6 +122,59 @@ static int get_device_serial_number(uint32_t dv_ind, char *serial, unsigned int
...
+    if (!hwloc_hide_errors()) {
+      const char *status_string;
+      rsmi_rc = rsmi_status_string(rsmi_rc, &status_string);
+      fprintf(stderr, "RSMI: GPU(%u): Failed to get serial number: %s\n", (unsigned)dv_ind, status_string);
+    }
+    return -1;
+  }
+  return 0;
+}
+
```
Comment (bgoglin) [inline hwloc/topology-rsmi.c:134]: Can you split the 3 above hunks to a separate commit sine it's unrelated to XGMI?

File: hwloc/topology-rsmi.c [hunk 2]
```diff
@@ -189,6 +254,26 @@ hwloc_rsmi_discover(struct hwloc_backend *backend, struct hwloc_disc_status *dst
...
+           (type == RSMI_IOLINK_TYPE_XGMI)) {
+        sprintf(buffer_ptr, "rsmi%u ", j);
+        buffer_ptr += 6;
+        if (buffer_ptr >= (buffer + sizeof(buffer)))
+          break;
+      }
+    }
+    if (buffer[0] != '\0')
+      hwloc_obj_add_info(osdev, "XGMIPeers", buffer);
+
```
Comment (bgoglin) [inline hwloc/topology-rsmi.c:276]: Can you do this instead?
```
char *buffer;
buffer = malloc(nb*15+1); /* "rsmi" + unsigned int + space = 15 chars max, + ending \0 */
if (buffer) {
 ....
 len = sprintf(...
 buffer_ptr += len;
 <remove the if break at end of loop>
...
free(buffer);
```

File: hwloc/topology-rsmi.c [hunk 3]
```diff
@@ -189,6 +254,26 @@ hwloc_rsmi_discover(struct hwloc_backend *backend, struct hwloc_disc_status *dst
     if (get_device_unique_id(i, buffer) == 0)
       hwloc_obj_add_info(osdev, "AMDUUID", buffer);
 
+    buffer[0] = '\0';
+    if (get_device_xgmi_hive_id(i, buffer) == 0)
+      hwloc_obj_add_info(osdev, "XGMIHIVEID", buffer);
```
Comment (bgoglin) [inline hwloc/topology-rsmi.c:259]: XGMIHiveID?

**Comment by miketxli** [comment]
I have requested the remove access for you a long time ago and still waiting for it. There are very limited spots available, in external network. I'm using a remove test machine, but it's in internal network. Let me push again. Will look the comments and make change.

File: hwloc/topology-rsmi.c (hunk 1 above)
Comment (miketxli) [reply hwloc/topology-rsmi.c:134]: OK

File: hwloc/topology-rsmi.c (hunk 3 above)
Comment (miketxli) [reply hwloc/topology-rsmi.c:259]: Ok

File: hwloc/topology-rsmi.c (hunk 2 above)
Comment (miketxli) [reply hwloc/topology-rsmi.c:276]: OK

**Comment by miketxli** [comment]
Have updated the code based on comments.

File: hwloc/topology-rsmi.c [hunk 4]
```diff
@@ -165,7 +213,9 @@ hwloc_rsmi_discover(struct hwloc_backend *backend, struct hwloc_disc_status *dst
   for (i=0; i<nb; i++) {
     uint64_t bdfid = 0;
     hwloc_obj_t osdev, parent;
-    char buffer[64];
+    char buffer[80];
```
Comment (bgoglin) [inline hwloc/topology-rsmi.c:216]: This change is unneeded now. Otherwise we're good, thanks a lot.
The other commits with unrelated changes can go in the same PR, I'll merge both at once.

Comment (bgoglin) [reply hwloc/topology-rsmi.c:216]: Can you rebase on top of latest master? Some tests in the CI require recent commits in master.

**Comment by miketxli** [comment]
Rebased and updated

**Comment by bgoglin** [comment]
Merged, thanks a lot !

**Comment by bgoglin** [comment]
@miketxli I asm asked whether we're going to expose memory information and maybe some stuff about the number of GPU tiles/cores whatever it's called. We expose those for CUDA, and we're going yo expose them for Intel OneAPI too. Do you have plans for these?

**Comment by miketxli** [comment]
Hi Brice, nice to hear from you. Yes, I do have plans for that. Let me talk to my manager and come back to you.
